File: BDFM.Hub/BDFM.Domain/Entities/Core/LeaveRequest.cs
using BDFM.Domain.Common;
using BDFM.Domain.Entities.Workflow;
using BDFM.Domain.Enums;
using System.ComponentModel.DataAnnotations;

namespace BDFM.Domain.Entities.Core
{
    public class LeaveRequest : AuditableEntity<Guid>
    {
        [StringLength(100)]
        public string EmployeeId { get; set; } = string.Empty; // معرف الموظف من نظام HR

        [StringLength(100)]
        public string? EmployeeNumber { get; set; } // رقم الموظف من HR

        [StringLength(255)]
        public string? EmployeeName { get; set; } // اسم الموظف من HR

        public Guid? OrganizationalUnitId { get; set; }
        public virtual OrganizationalUnit? OrganizationalUnit { get; set; }

        public Guid? CreatedByUserId { get; set; }
        public virtual User? CreatedByUser { get; set; }

        public LeaveTypeEnum LeaveType { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public int RequestedDays { get; set; }  // عدد الأيام المطلوبة
        public int? ApprovedDays { get; set; }  // عدد الأيام الموافق عليها
        public decimal TotalBalance { get; set; } = 0; // الرصيد الكلي من نظام HR
        public decimal MonthlyBalance { get; set; } = 0; // الرصيد الشهري - 3 أيام للعادية
        public decimal UsedBalance { get; set; } = 0; // المستخدم
        public decimal AvailableBalance { get; set; } = 0; // المتاح = Total - Used
        public decimal MonthlyUsedBalance { get; set; } = 0; // المستخدم شهرياً

        public DateTime? LastMonthlyResetDate { get; set; } // تاريخ آخر إعادة تعيين شهري
        public LeaveRequestStatusEnum Status { get; set; }

        public string? Reason { get; set; } // Text type in DB // السبب في الطلب
        public string? RejectionReason { get; set; } // Text type in DB // السبب في الرفض 

        public DateTime? ApprovedAt { get; set; } // تاريخ الموافقة
        public Guid? ApprovedByUserId { get; set; } // معرف الموافق عليه
        public virtual User? ApprovedByUser { get; set; }

        public Guid? CancelledByUserId { get; set; } // معرف الملغي
        public virtual User? CancelledByUser { get; set; }
        public DateTime? CancelledAt { get; set; } // تاريخ الملغي
        public string? CancellationReason { get; set; } // Text type in DB // السبب في الملغي

        public bool IsInterrupted { get; set; } = false; // هل تم القطع
        public DateTime? ActualEndDate { get; set; } // تاريخ النهاية الفعلية

        [StringLength(100)]
        public string? RequestNumber { get; set; } // مثل "LEAVE-2025-001"

        // Navigation Properties
        public virtual ICollection<LeaveInterruption> Interruptions { get; set; } = new HashSet<LeaveInterruption>();
        public virtual ICollection<LeaveCancellation> Cancellations { get; set; } = new HashSet<LeaveCancellation>();
        public virtual ICollection<LeaveBalanceHistory> BalanceHistories { get; set; } = new HashSet<LeaveBalanceHistory>();
        public virtual ICollection<WorkflowStep> WorkflowSteps { get; set; } = new HashSet<WorkflowStep>();
    }
}

File: frontend-app/src/features/leave-request/types/leave-request.ts
import { LeaveInterruption } from '@/features/leave-interruption/types/leave-interruption';
import { LeaveCancellation } from '@/features/leave-cancellation/types/leave-cancellation';

export interface LeaveRequest {
  id?: string;
  employeeId?: string;
  employeeNumber?: string;
  employeeName?: string;
  organizationalUnitId?: string;
  organizationalUnitName?: string;
  createdByUserId?: string;
  createdByUserName?: string;
  leaveType?: number;
  leaveTypeName?: string;
  startDate?: string;
  endDate?: string;
  requestedDays?: number;
  approvedDays?: number;
  status?: number;
  statusName?: string;
  reason?: string;
  rejectionReason?: string;
  approvedAt?: string;
  approvedByUserId?: string;
  approvedByUserName?: string;
  cancelledByUserId?: string;
  cancelledByUserName?: string;
  cancelledAt?: string;
  cancellationReason?: string;
  isInterrupted?: boolean;
  actualEndDate?: string;
  requestNumber?: string;
  createAt?: string;
  createBy?: string;
  lastUpdateAt?: string;
  lastUpdateBy?: string;
  statusId?: number;
  interruptions?: LeaveInterruption[];
  cancellations?: LeaveCancellation[];
}

export interface LeaveRequestList {
  id?: string;
  employeeId?: string;
  employeeNumber?: string;
  employeeName?: string;
  organizationalUnitId?: string;
  organizationalUnitName?: string;
  leaveType?: number;
  leaveTypeName?: string;
  startDate?: string;
  endDate?: string;
  requestedDays?: number;
  approvedDays?: number;
  status?: number;
  statusName?: string;
  requestNumber?: string;
  isInterrupted?: boolean;
  approvedAt?: string;
  approvedByUserName?: string;
  createAt?: string;
  statusId?: number;
}

export interface CreateLeaveRequestPayload {
  employeeId: string;
  organizationalUnitId?: string;
  leaveType: number;
  startDate: string;
  endDate: string;
  reason?: string;
}

export interface UpdateLeaveRequestPayload {
  id: string;
  leaveType?: number;
  startDate?: string;
  endDate?: string;
  reason?: string;
}

export interface ApproveLeaveRequestPayload {
  id: string;
  approvedDays?: number;
  notes?: string;
}

export interface RejectLeaveRequestPayload {
  id: string;
  rejectionReason: string;
}

export interface CancelLeaveRequestPayload {
  id: string;
  cancellationReason?: string;
}

export interface InterruptLeaveRequestPayload {
  id: string;
  interruptionDate: string;
  returnDate: string;
  interruptionType: number;
  reason?: string;
}

export interface GetAllLeaveRequestsQuery {
  page?: number;
  pageSize?: number;
  searchText?: string;
  employeeId?: string;
  organizationalUnitId?: string;
  leaveType?: number;
  status?: number;
  startDateFrom?: string;
  startDateTo?: string;
}

export interface GetLeaveRequestsByEmployeeIdQuery {
  page?: number;
  pageSize?: number;
  employeeId?: string;
  leaveType?: number;
  status?: number;
}

export interface GetLeaveRequestsByStatusQuery {
  page?: number;
  pageSize?: number;
  status?: number;
  organizationalUnitId?: string;
  employeeId?: string;
}

export enum LeaveRequestStatus {
  Draft = 1,
  PendingApproval = 2,
  Approved = 3,
  Rejected = 4,
  Cancelled = 5,
  Interrupted = 6,
  Completed = 7
}

export const LeaveRequestStatusDisplay: Record<LeaveRequestStatus, string> = {
  [LeaveRequestStatus.Draft]: 'مسودة',
  [LeaveRequestStatus.PendingApproval]: 'قيد الموافقة',
  [LeaveRequestStatus.Approved]: 'موافق عليه',
  [LeaveRequestStatus.Rejected]: 'مرفوض',
  [LeaveRequestStatus.Cancelled]: 'ملغي',
  [LeaveRequestStatus.Interrupted]: 'مقطوع',
  [LeaveRequestStatus.Completed]: 'منتهية'
};

export enum LeaveType {
  RegularDaily = 1,
  LongTerm = 2,
  Annual = 3,
  Study = 4,
  Maternity = 5,
  Mourning = 6,
  Sick = 7
}

export const LeaveTypeDisplay: Record<LeaveType, string> = {
  [LeaveType.RegularDaily]: 'عادية/يومية',
  [LeaveType.LongTerm]: 'طويلة المدة',
  [LeaveType.Annual]: 'سنوية',
  [LeaveType.Study]: 'دراسية',
  [LeaveType.Maternity]: 'أمومة',
  [LeaveType.Mourning]: 'عدة',
  [LeaveType.Sick]: 'مرضية'
};

